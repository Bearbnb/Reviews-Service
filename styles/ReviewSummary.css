span:before {
   content: "\2605";
   position: absolute;
}
.break {
    margin-top: 16px;
    margin-bottom: 16px;
    border-bottom: 1px solid #EBEBEB;
    box-sizing: border-box

}
.textHeader {
    display: table;
}
.reviewTitle {
    display: table-cell;
}
.headerStars {
    display: inline-block; 
    margin-left: 16px;
}