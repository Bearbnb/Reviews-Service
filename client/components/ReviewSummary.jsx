import React from 'react';
import ReactDOM from 'react-dom';
import ReviewSummaryLine from './ReviewSummaryLine.jsx'
import Search from './Search.jsx'

import styles from '../../styles/ReviewSummary.css';

class ReviewSummary extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            termSearched:'',
            reviewTotal: 1,
            average: 0,
            categories: [{Accuracy: 0}, 
            {Comunication: 0}, 
            {Cleanliness: 0}, 
            {Location: 0}, 
            {CheckIn: 0}, 
            {Value: 0}]
        };
    }

    componentDidUpdate(prevProps) {
        if (this.props.reviews !== prevProps.reviews) {
            this.setState({
                reviewTotal: this.props.reviews.length,
                categories : this.getAverages(), 
                average: this.calculateAverage(this.getAverages())
            })
        }
    }
    calculateAverage(arr) {
        let total = 0;
        arr.forEach(review => {
            let description = Object.keys(review);
            let rating = review[description[0]];
            total += rating;
        });
        total = total / arr.length;
        return Math.round(total)
    }
    
    getAverages(){
        let vals = {accuracy: 0,
            communication: 0,
            cleanliness: 0,
            location: 0,
            checkin: 0,
            value: 0
        }
        this.props.reviews.forEach(review =>{
            review.rating.forEach(rate =>{
                let description = Object.keys(rate);
                let rating = rate[description[0]];
                vals[description]+=rating
                
            })
        })
        let newCategories = [];
        for(const key in vals ){
            let value = (vals[key] / this.props.reviews.length).toFixed(2)
            newCategories.push({[key]: Number(value)});
        }
       return newCategories;
    }
    handleSearch(e){
        this.props.onClick(e);
        this.setState({
            termSearched: e
        })
    }

    render() {
        var stars = ['☆', '☆', '☆', '☆', '☆'].map(star =>
            <span className={styles.reviewStar}></span>)
        if(this.props.searched===false)
        {
        return (
            <div className = {styles.averageContainer}>
                <div className={styles.break}>
                </div>
                <div className={styles.averageHeader}>
                    <div className = {styles.textHeader}>
                    <h2 className = {styles.reviewTitle}>{this.props.reviews.length} Reviews </h2>
                       <div className ={styles.headerStars}>
                        {stars} 
                       </div> 
                    </div>
                    <div className = {styles.search}>
                        <Search onClick={this.handleSearch.bind(this)} searched={this.props.searched}/>
                    </div>
                </div>
                <div className = {styles.break}>
                </div>
                <div className = {styles.breakdownComponent} >
                    {this.state.categories.map((category) =>
                        <ReviewSummaryLine category={category}/>
                    )} 
                   
                </div>
            </div>
        )} else {
            return (
                <div className={styles.averageContainer}>
                    <div className={styles.break}>
                    </div>
                    <div className={styles.averageHeader}>
                        <div className={styles.textHeader}>
                            <h2 className={styles.reviewTitle}>{this.props.reviews.length} Reviews
                       <div className={styles.headerStars}>
                                    {stars}
                                </div>
                            </h2>
                        </div>
                        <div className={styles.search}>
                            <Search onClick={this.props.onClick} searched = {this.props.searched}/>
                        </div>
                    </div>
                    <div>
                        <h4>{this.props.reviews.length} guests have mentioned "{this.state.termSearched}"</h4>
                        <button id ={styles.resetReviews} onClick={()=>this.props.resetHomes()}> Back to all reviews </button>
                    </div>
                    <div className={styles.break}>
                    </div>
                </div>
            )
        }
    }
}

export default ReviewSummary;


