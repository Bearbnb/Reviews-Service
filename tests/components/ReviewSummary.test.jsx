import { shallow, mount, render } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import ReviewSummary from '../../client/components/ReviewSummary';

describe('<ReviewSummary />', () => {
  it('should render The summary of all reviews to the DOM', () => {
    const wrapper = shallow(<ReviewSummary
      reviews={[]}
      searched={false}
      resetHomes={jest.fn()}
      length={0}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render something else after searching', () => {
    const wrapper = shallow(<ReviewSummary
      reviews={[]}
      searched
      resetHomes={jest.fn()}
      length={0}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
