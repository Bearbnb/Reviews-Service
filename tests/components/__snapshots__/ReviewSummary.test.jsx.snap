// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReviewSummary /> should render The summary of all reviews to the DOM 1`] = `
<div
  className="averageContainer"
>
  <div
    className="break"
  />
  <Header
    average={0}
    length={0}
    onClick={[Function]}
    ratingToStars={[Function]}
    searched={false}
  />
  <div
    className="break"
  />
  <SubHeader
    categories={
      Array [
        Object {
          "Accuracy": 0,
        },
        Object {
          "Comunication": 0,
        },
        Object {
          "Cleanliness": 0,
        },
        Object {
          "Location": 0,
        },
        Object {
          "CheckIn": 0,
        },
        Object {
          "Value": 0,
        },
      ]
    }
    ratingToStars={[Function]}
    resetHomes={[MockFunction]}
    reviews={Array []}
    searched={false}
    term=""
  />
</div>
`;

exports[`<ReviewSummary /> should render something else after searching 1`] = `
<div
  className="averageContainer"
>
  <div
    className="break"
  />
  <Header
    average={0}
    length={0}
    onClick={[Function]}
    ratingToStars={[Function]}
    searched={true}
  />
  <SubHeader
    categories={
      Array [
        Object {
          "Accuracy": 0,
        },
        Object {
          "Comunication": 0,
        },
        Object {
          "Cleanliness": 0,
        },
        Object {
          "Location": 0,
        },
        Object {
          "CheckIn": 0,
        },
        Object {
          "Value": 0,
        },
      ]
    }
    ratingToStars={[Function]}
    resetHomes={[MockFunction]}
    reviews={Array []}
    searched={true}
    term=""
  />
  <div
    className="break"
  />
</div>
`;
